release-23.1: roachtest: selective test as a part of nightly run

nameisbhaskar opened the pull request:
Backport 1/1 commits from #125481.

/cc @cockroachdb/release

---
Currently, we are running all the tests for this release. This causes the pipeline to run for a very long time.
Also, running all the tests everyday incurs cost.
So, with this change, we selectively run the tests based on the following criteria:
1. Tests that have failed at least once in last 30 days.
2. Tests not run in last 7 days.
3. Tests that are not older than 20 days.
4. 30% of the tests that have been successfully running.

Note that this change does not have the changes in master like considering the preempt failures as success as spot VM change is not part of this release.

Informs: #119630
Epic: none
Release justification: test engineering changes

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/128068)
<!-- Reviewable:end -->

nameisbhaskar committed:
Currently, we are running all the tests for this release. This causes the pipeline to run for a very long time.
Also, running all the tests everyday incurs cost.
So, with this change, we selectively run the tests based on the following criteria:

Tests that have failed at least once in last 30 days.
Tests not run in last 7 days.
Tests that are not older than 20 days.
30% of the tests that have been successfully running.
Note that this change does not have the changes in master like considering the preempt failures as success as spot VM change is not part of this release.

Informs: #119630
Epic: none
Release justification: test engineering changes

srosenberg reviewed (approved):
Note, we also need to backport the following PRs into this branch, as well as, 23.2 and 24.1; [1] has already been backported into 24.2, but not [2], so that as well.

[1] https://github.com/cockroachdb/cockroach/pull/127974
[2] https://github.com/cockroachdb/cockroach/pull/127752

nameisbhaskar commented:
Thanks @srosenberg for reviewing this. I will take up the backporting of the PRs immediately after merging this.
